[PIR-1311]-fix for found regression

PashkovichAndrei opened the pull request.

PashkovichAndrei commented:
@nantunes @dcleao could you please have a look?

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
mvn -Dsurefire.runOrder=alphabetical -B -fn -f 'pom.xml' -pl 'impl/client' -P '!assemblies' -amd clean install && mvn -B -f 'pom.xml' -pl 'impl/client' -P '!assemblies' -amd site
```

#### Cleanup Commands
```sh
mvn -B -f 'pom.xml' -pl 'impl/client' -P '!assemblies' -amd build-helper:remove-project-artifact
```

#### Changed files

```console
  impl/client/src/main/javascript/web/dojo/pentaho/common/FilterDialog.js
```

dcleao commented on impl/client/src/main/javascript/web/dojo/pentaho/common/FilterDialog.js:
```diff
@@ -420,7 +420,9 @@ define(["dojo/_base/declare", "dijit/_WidgetBase", "dijit/_TemplatedMixin", "dij
       str.forEach(function(val) {
-        val.replace(/&amp;/g, '&');
+        if (typeof str === 'string' || str instanceof String) {
+          val.replace(/&amp;/g, '&');
```
Unless `val` is actually a `String`, not common, replacing characters on a string returns a new string and thus have no side effect on the given array.
So I think you should do `str[index] = val.replace( ... );`, where `index` is the second argument to the `forEach`.

The same applies to the replace line below, when not in the forEach case.
Did this ever work? Are we using `String` objects — which is usually not good practice — for this to  work?

PashkovichAndrei commented:
@dcleao, this is a found regression: ![image](https://user-images.githubusercontent.com/24562938/33420852-3faa0b3a-d5c1-11e7-8b5a-e9b93526413b.png)
I add one more if statement into forEach method to avoid NPEs. Except check only for null, I've added check for String objects & string literals: `typeof str === 'string' || str instanceof String`; this check will be true only if we pass String object or string literal into it.

Also I've reworked PR regarding your comment above. Could you please do review?

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
mvn -Dsurefire.runOrder=alphabetical -B -fn -f 'pom.xml' -pl 'impl/client' -P '!assemblies' -amd clean install && mvn -B -f 'pom.xml' -pl 'impl/client' -P '!assemblies' -amd site
```

#### Cleanup Commands
```sh
mvn -B -f 'pom.xml' -pl 'impl/client' -P '!assemblies' -amd build-helper:remove-project-artifact
```

#### Changed files

```console
  impl/client/src/main/javascript/web/dojo/pentaho/common/FilterDialog.js
```

dcleao commented on impl/client/src/main/javascript/web/dojo/pentaho/common/FilterDialog.js:
```diff
@@ -419,11 +419,13 @@ define(["dojo/_base/declare", "dijit/_WidgetBase", "dijit/_TemplatedMixin", "dij
-      str.forEach(function(val) {
-        val.replace(/&amp;/g, '&');
+      str.forEach(function(val, index) {
+        if (typeof str === 'string' || str instanceof String) {
```
Should it be: `if (typeof val === 'string' || val instanceof String)`
You should start testing your changes...

PashkovichAndrei replied:
Sorry, my bad...
